src/colorizer/ColorizeCanvas.ts

```diff
@@ -486,21 +495,25 @@ export default class ColorizeCanvas {
     // feature thresholds or not.
     // TODO: Optimize memory by using a true boolean array? Bit-level manipulation to fit it within Uint8Array?
     // TODO: If optimizing, use fuse operation via shader.
-    const inRangeIds = new Uint8Array(this.dataset.numObjects);
-    inRangeIds.fill(1);
-
-    for (const threshold of thresholds) {
-      const featureData = this.dataset.getFeatureData(threshold.featureName);
-      // Ignore thresholds with features that don't exist in this dataset or whose units don't match
-      if (!featureData || featureData.units !== threshold.units) {
-        continue;
-      }
-      for (let i = 0, n = inRangeIds.length; i < n; i++) {
-        if (inRangeIds[i] === 1 && (featureData.data[i] < threshold.min || featureData.data[i] > threshold.max)) {
-          inRangeIds[i] = 0;
+    let inRangeIds = new Uint8Array(this.dataset.numObjects);
+
+    // Ignore thresholds with features that don't exist in this dataset or whose units don't match
+    const validThresholds = thresholds.filter((threshold) => {
+      const featureData = this.dataset?.tryGetFeatureData(threshold.featureName);
+      return featureData && featureData.units === threshold.units;
+    });
+
+    inRangeIds.forEach((_value, id, arr) => {
+      for (const threshold of validThresholds) {
+        const featureData = this.dataset?.tryGetFeatureData(threshold.featureName);
+        if (featureData && !this.isValueWithinThreshold(featureData.data[id], threshold)) {
+          arr[id] = 0;
+          return;
         }
       }
-    }
+      arr[id] = 1;
+    });
```

Comment:
This loop gives better performance in the best-case scenario (we can iterate over each `id` once instead of minimum `id * num_thresholds`) and is a little easier to read IMO.